packages/docprovider/src/notebookCellExecutor.ts

```diff
@@ -0,0 +1,168 @@
+  private _serverSettings: ServerConnection.ISettings;
+
+  /**
+   * Constructor
+   *
+   * @param options Constructor options; the contents manager, the collaborative drive and optionally the server settings.
+   */
+  constructor(options: {
+    contents: Contents.IManager;
+    drive: ICollaborativeDrive;
+    serverSettings?: ServerConnection.ISettings;
+  }) {
+    this._contents = options.contents;
+    this._drive = options.drive;
+    this._serverSettings =
+      options.serverSettings ?? ServerConnection.makeSettings();
+  }
+
+  /**
+   * Execute a given cell of the notebook.
+   *
+   * @param options Execution options
+   * @returns Execution success status
+   */
+  async runCell({
+    cell,
+    notebook,
+    notebookConfig,
+    onCellExecuted,
+    onCellExecutionScheduled,
+    sessionContext,
+    sessionDialogs,
+    translator
+  }: INotebookCellExecutor.IRunCellOptions): Promise<boolean> {
+    translator = translator ?? nullTranslator;
+    const trans = translator.load('jupyterlab');
+
+    switch (cell.model.type) {
+      case 'markdown':
+        (cell as MarkdownCell).rendered = true;
+        cell.inputHidden = false;
+        onCellExecuted({ cell, success: true });
+        break;
+      case 'code':
+        if (sessionContext) {
+          if (sessionContext.isTerminating) {
+            await showDialog({
+              title: trans.__('Kernel Terminating'),
+              body: trans.__(
+                'The kernel for %1 appears to be terminating. You can not run any cell for now.',
+                sessionContext.session?.path
+              ),
+              buttons: [Dialog.okButton()]
+            });
+            break;
+          }
+          if (sessionContext.pendingInput) {
+            await showDialog({
+              title: trans.__('Cell not executed due to pending input'),
+              body: trans.__(
+                'The cell has not been executed to avoid kernel deadlock as there is another pending input! Submit your pending input and try again.'
+              ),
+              buttons: [Dialog.okButton()]
+            });
+            return false;
+          }
+          if (sessionContext.hasNoKernel) {
+            const shouldSelect = await sessionContext.startKernel();
+            if (shouldSelect && sessionDialogs) {
+              await sessionDialogs.selectKernel(sessionContext);
+            }
+          }
+
+          if (sessionContext.hasNoKernel) {
+            cell.model.sharedModel.transact(() => {
+              (cell.model as ICodeCellModel).clearExecution();
+            });
+            return true;
+          }
+
+          const kernelId = sessionContext?.session?.kernel?.id;
+          const apiURL = URLExt.join(
+            this._serverSettings.baseUrl,
+            `api/kernels/${kernelId}/execute`
+          );
+          const cellId = cell.model.sharedModel.getId();
+
+          // jupyverse case - it is undefined in jupyter-server
```

Comment:
Maybe we should make it part of the document state? I opened:
- https://github.com/jupyter-server/jupyter_ydoc/pull/198
- https://github.com/jupyterlab/jupyter-collaboration/pull/217

Reply:
This could be a more efficient way than using a cache as proposed in https://github.com/jupyterlab/jupyter-collaboration/pull/307#discussion_r1598031145